pkg/webhook/server.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"path"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/cert"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/cert/writer"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/types"
+)
+
+// ServerOptions are options for configuring an admission webhook server.
+type ServerOptions struct {
+	// KVMap can be used to pass in any information that the admission.Handler needs.
+	KVMap map[string]interface{}
+
+	// Port is the port number that the server will serve. Default
+	Port int32
+
+	// CertDir is the directory that contains the server key and certificate
+	CertDir string
+
+	// Client is a kubernetes client that knows how to talk to a kubernetes cluster.
+	// Client will be injected by the manager if not set.
+	Client client.Client
+
+	// BootstrapOptions contains the options for bootstrapping the admission server.
+	BootstrapOptions
```

Comment:
is this bootstrap stuff optional?  What if I don't want to run boostrap, and want to create the objects myself?

Reply:
Yes, it's optional.
We also provide a dryrun mode that generates the resources in yaml instead of apply them to the server.

Reply:
it should probably be a pointer then